assignment5/q1.cpp

```diff
@@ -0,0 +1,184 @@
+
+        return;
+    }
+
+    //make connections
+    for(int word_id = 0; word_id < dictionary.size()-1; word_id++) {
+        string word1 = dictionary[word_id];
+        string word2 = dictionary[word_id + 1];
+        int char_id = 0;
+        while(word1[char_id] == word2[char_id])
+            char_id++;
+
+        if(connections[word1[char_id]] == NULL) {
+            letters.insert(word1[char_id]);
+            in_number[word1[char_id]] = 0;
+            connections[word1[char_id]] = &(*new vector<char>);
+        }
+        if(connections[word2[char_id]] == NULL) {
+            letters.insert(word2[char_id]);
+            in_number[word2[char_id]] = 0;
+            connections[word2[char_id]] = &(*new vector<char>);
+        }
+        //add the directed edge
+        (*connections[word1[char_id]]).push_back(word2[char_id]);
+        letters.erase(word2[char_id]);
+        in_number[word2[char_id]]++;
+
+    }
+
+    return;
+
+}
+
+vector<char> make_alphabet_from_rules(vector<char> *connections[256], set<char> &letters, vector<int> &in_number) {
+
+    int current_id = 0;
+
+    vector<char> alphabet;
+    set<char>::iterator current_letter = letters.begin();
+
+    //if dictionary was empty
+    if(letters.empty()) {
+        cout<<"empty"<<endl;
+        return alphabet;
+    }
+
+    //while there are still letters to process
+    while(!letters.empty()) {
+
+        alphabet.push_back(*current_letter);
+
+        //if dictionary had one word consisting of more than one char
+        if(connections[*current_letter] == NULL){
+            cout<<"cannot make rules"<<endl;
+            return alphabet;
+        }
+
+        vector<char> *vector_for_letter = &(*(connections[*current_letter]));
+
+        //look for all connections from letter
+        for(int vertex = 0; vertex < (*vector_for_letter).size(); vertex++) {
+            char vertex_letter = (*vector_for_letter)[vertex];
+            in_number[vertex_letter]--;
+            if(in_number[vertex_letter] == 0) {
+                letters.insert(vertex_letter);
+            }
+        }
+        //delete letter's connections
+        vector<char> *row = &(*(connections[*current_letter]));
+        letters.erase(*current_letter);
+        (*row).erase((*row).begin(), (*row).end());
+        current_letter = letters.begin();
+
+    }
+
+    //if there are still any connections
+    for(int i = 0; i < 256; i++){
+        if((&(*connections[i]) != NULL) && (*connections[i]).empty() != 1) {
+            //if cycle or contradictory rules
+            cout<<"cannot make rules"<<endl;
+            alphabet.erase(alphabet.begin(), alphabet.end());
+            return alphabet;
+        }
+    }
+
+    //print out alphabet
+    for(int i = 0; i < alphabet.size(); i++) {
+        cout<<alphabet[i]<<" ";
+    }
+    cout<<endl;
+
+    return alphabet;
+
+}
+
+vector<char> create_alphabet_from_dictionary(vector<string> dictionary) {
+
+    set<char> letters;
+    vector<char> *rules[256];
```

Comment:
An array of pointers to vectors... I find this very hard for the reader to follow.
is `rules[5]` a vector? a pointer to a vector? Can I expect it to have value or it is NULL? If it is NULL, what does it mean?

I suggest a few alternatives:
First of all, a vector is almost always better than an array. A vector of pointers is possible, initialized to `NULL`:
`vector<vector<char>*> rules(256, NULL);`
This also saves you the initialization loop inside `make_rules`.

Furthermore - what does `NULL` indicate here? It says that this value is not used, that this letter does not exist in the dictionary. In such cases, you can use a `map` to keep the mapping from a letter to its neighbors, but only for the letters you want:
`map<char, vector<char>> rules;`
Now it is clear that this maps a character to a vector of other characters. And you can keep a very clear syntax: `rules['a'].push_back('b');`
And there is never a danger of hitting a null pointer!